### PR TITLE
review: fix: getTypeDeclaration() on nested record type

### PR DESCRIPTION
**Pull request by I-Al-Istannen**
The nested visitor did not consider records as it was written before their existence.

We might want to find a better upgrade story for such cases.

### CODE REVIEW

**Review by MartinWitt — APPROVED** [review]

**Comment by MartinWitt** [comment]
@I-Al-Istannen @SirYwell could both of you also mark this PR for approval (after a review) and we simply merge it?

**Comment by I-Al-Istannen** [comment]
Github won't let me approve it as I am the author. But consider this my approval :P

**Review by SirYwell — DISMISSED** [review]

**Comment by MartinWitt** [comment]
@monperrus could you have a brief look and merge this PR? As all of us three are (co)authors.

**Review by slarse — CHANGES_REQUESTED** [review]
Looks functionally correct, I have a couple of minor comments

File: src/test/java/spoon/test/ctType/CtTypeTest.java [hunk 1]
```diff
@@ -305,4 +306,26 @@ public void testEnumPackage() {
...
+		launcher.getEnvironment().setComplianceLevel(17);
+		CtType<?> foo = launcher.buildModel().getAllTypes().iterator().next();
+		assertEquals(foo, foo.getReference().getTypeDeclaration());
+
+		for (CtType<?> nestedType : foo.getNestedTypes()) {
+			assertEquals(nestedType, nestedType.getReference().getTypeDeclaration());
+			for (CtType<?> type : nestedType.getNestedTypes()) {
+				assertEquals(type, type.getReference().getTypeDeclaration());
+			}
+		}
```
[the remark this answers is not in the document]
Comment (I-Al-Istannen) [reply src/test/java/spoon/test/ctType/CtTypeTest.java]: Yea, very good point. I have broken a few testsuites in the past without noticing any failure… `∀` for an empty set isn't as meaningful as you would hope :P

I opted for an adaption of `b)` which should behave identically. I hope that's okay.

File: src/main/java/spoon/support/reflect/declaration/CtTypeImpl.java [hunk 2]
```diff
@@ -357,49 +362,53 @@ private boolean checkType(CtType<?> type) {
...
+
+			@Override
+			public void visitCtTypeParameter(CtTypeParameter typeParameter) {
+				// ignored, can not harbour any nested types
+			}
+
+			@Override
+			protected void enter(CtElement e) {
+				if (e instanceof CtType) {
+					throw new SpoonException("Unknown type, extend scanner: " + e.getClass());
```
[the remark this answers is not in the document]
Comment (I-Al-Istannen) [reply src/main/java/spoon/support/reflect/declaration/CtTypeImpl.java]: Good point, I have changed the wording. Is it fine with you or do you want further changes to it :)

**Review by slarse — APPROVED** [review]
LGTM, thanks everyone. Now to see if I can manage to merge without removing all coauthorships.

File: src/main/java/spoon/support/reflect/declaration/CtTypeImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (slarse) [reply src/main/java/spoon/support/reflect/declaration/CtTypeImpl.java]: Looks good!

File: src/test/java/spoon/test/ctType/CtTypeTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (slarse) [reply src/test/java/spoon/test/ctType/CtTypeTest.java]: :+1:

**Comment by monperrus** [comment]
thanks all!

**Pull request merged**